core/domain/stats_jobs_one_off.py

```diff
@@ -1694,3 +1694,159 @@ def reduce(key, values):
+            yield (
+                FillExplorationStatsOneOffJob.NO_STATS_MODELS_AVAILABLE, exp_id)
+            return
+
+        exp_list = exp_fetchers.get_multiple_explorations_by_version(
+            latest_exp_model.id, exp_versions)
+
+        change_lists = [
+            [
+                exp_domain.ExplorationChange(commit_cmd)
+                for commit_cmd in snapshot['commit_cmds']
+            ]
+            for snapshot in exp_models.ExplorationModel.get_snapshots_metadata(
+                latest_exp_model.id,
+                python_utils.RANGE(1, latest_exp_model.version + 1))
+        ]
+        missing_exp_stats = []
+        missing_state_stats = []
+
+        zipped_items = list(
+            python_utils.ZIP(exp_stats_list, exp_list, change_lists))
+        revert_commit_cmd = exp_models.ExplorationModel.CMD_REVERT_COMMIT
+        for i, (exp_stats, exp, change_list) in enumerate(zipped_items):
+            revert_cmd = python_utils.NEXT(
+                (
+                    change for change in change_list
+                    if change.cmd == revert_commit_cmd
+                ), None)
+            revert_to_version = revert_cmd and revert_cmd.version_number
+            new_exp_version = None
+
+            if revert_to_version is not None:
+                if revert_to_version == 1:
+                    yield (
+                        FillExplorationStatsOneOffJob.V1_EXPLORATION_MODEL,
+                        exp.id)
+                    return
+                exp_versions_diff = None
+                prev_exp_stats = exp_stats_list[revert_to_version - 2]
+                prev_exp = exp_list[revert_to_version - 2]
+                new_exp_version = revert_to_version
+            else:
+                if exp.version == 1:
+                    yield (
+                        FillExplorationStatsOneOffJob.V1_EXPLORATION_MODEL,
+                        exp.id)
+                    return
+                exp_versions_diff = exp_domain.ExplorationVersionsDiff(
+                    change_list)
+                prev_exp_stats = exp_stats_list[exp.version - 2]
+                prev_exp = exp_list[exp.version - 2]
+                new_exp_version = exp.version
+
+            # Fill missing Exploration-level stats.
+            if exp_stats is not None:
+                num_valid_exp_stats += 1
+            else:
+                exp_stats = prev_exp_stats and prev_exp_stats.clone()
+
+                if exp_versions_diff is not None:
+                    exp_stats = stats_services.advance_version_of_exp_stats(
+                        new_exp_version, exp_versions_diff, exp_stats, None,
+                        None)
+                else:
+                    exp_stats.version = exp.version
+                exp_stats_list[i] = exp_stats
+                missing_exp_stats.append(
+                    'Exp id: %s, Version: %s' % (exp_id, exp.version))
+                new_exploration_stats = (
```

Comment:
Why are we creating new empty stats if we already spent the work cloning/advancing a previous version? Let's just save it here and filter out `None` values later?